[FIX] Authorship for Proposals list have disappeared

tramuntanal opened the pull request:
#### :tophat: What? Why?
Make proposal m-card render its authorship again.

#### :pushpin: Related Issues
- Fixes #3725 

#### :clipboard: Subtasks
- [x] Add `CHANGELOG` entry
- [-] Add documentation regarding the feature 
- [x] Add/modify seeds
- [x] Add tests
- [x] Apply fix

### :camera: Screenshots (optional)
![Description](URL)

mrcasals commented on decidim-core/app/cells/decidim/card_m/header.erb:
```diff
@@ -14,4 +14,5 @@
   <% end %>
 
   <%= render :author if has_author? %>
+  <%= cell("decidim/coauthorships", model, has_actions: has_actions?) if coauthorable? %>
```
Can a resource be both `Authorable` and `Coauthorable`?  If they can't, then I'd join lines 16 and 17 and make the cell keep the logic on how to render the authors.

mrcasals replied:
Something like:

```ruby
def author
  return render :author if has_author?
  return cell("decidim/coauthorships", model, has_actions: has_actions?) if coauthorable?
  nil
end
```

And then on the views replace lines 16-17 with `<%= author %>`

agustibr replied:
@mrcasals I agree

mrcasals commented on decidim-core/app/cells/decidim/coauthorships_cell.rb:
```diff
@@ -0,0 +1,44 @@
+    include Decidim::ApplicationHelper
+
+    def show
+      if model.respond_to?(:official?) && model.official?
```
`if model.try(:official?)`

tramuntanal replied:
I don't like to use `try` here because it suggests that the model should have this method but the model can be nil.

The longer expression expresses best the duck typing I guess.

mrcasals replied:
👍

tramuntanal replied:
I've extracted the logic into a method.

mrcasals commented on decidim-core/app/cells/decidim/coauthorships_cell.rb:
```diff
@@ -0,0 +1,44 @@
+
+    def has_actions?
+      return false if options[:has_actions] == false
+      true
```
You can replace this whole method with `options[:has_actions] == true`

mrcasals commented on decidim-proposals/lib/decidim/proposals/component.rb:
```diff
@@ -151,7 +149,12 @@
         published_at: Time.current
       )
-      proposal.add_coauthor(author, user_group: user_group) if n.even?
+      if n.positive?
```
Won't `n` be always positive? Why not keep the previous `n.even`?

tramuntanal replied:
First `n` will be zero: for officialized proposals. Then each proposal will have `n` coauthors.

This way we will have proposals with "all" coauthorships: official, 1 author, many coauthors.

mrcasals replied:
Oh, I see! Nice!

mrcasals commented on decidim-core/app/cells/decidim/card_m_cell.rb:
```diff
@@ -58,18 +58,6 @@ def decidim
       Decidim::Core::Engine.routes.url_helpers
     end
 
-    def has_author?
```
I think we should keep the logic here, so each card can overwrite how to render their author if needed

https://github.com/decidim/decidim/pull/3727#discussion_r198733600

tramuntanal replied:
Cells are going complicated and what you propose is breaking encapsulation. Why a parent card should decide how authorships should be rendered if we already have cells to tackle this visualization details?

If we want a coherent UI we need to promote homogeneization, not the contrary, don't you think?
Also YAGNI, if some card need to overwrite this authorship cell for some reason, it will be better to tackle the problem when we know the whole information. In any case, we can extract the call to the coauthorships cell into a method so that it can be overwritten with specific logic.

mrcasals replied:
Ok, then, but please check that no other card is overwriting the `author.erb` view then 😄

tramuntanal replied:
I guess I misunderstood your first comment! I first thought you meant to keep the logic of selecting what to render (if author cell or coauthorship cell) but I think you meant to have a way to select if authors should be rendered or not.

Then I think you are right, to keep a has_authors? method to flag if authorship should be rendered or not. Is this what you meant?

mrcasals replied:
No, you understood right! 😄 But you made a good point, so 👍 for the change.

Still, what you propose (the `has_authors?` flag) looks very useful, so 👍 on adding it!

tramuntanal commented:
Hi @decidim/lot-core , it seems after merging master into this branch some kind of decidim version inconsistency appeard.
Locally everything runs ok, but in circle I'm having issues between 0.12 and 0.14.

I see there are no more commits in master. Are you working on this?

mrcasals commented:
@tramuntanal `master` should be fixed right now, please rebase/update your branch with `master` to run the tests 😄

tramuntanal commented:
Ready to be merged @decidim/lot-core

mrcasals commented:
@tramuntanal if you fix the CHANGELOG we'll merge this skipping the CI 😄 Code looks good!

tramuntanal commented:
I've just updated the CHANGELOG.md I don't know what happened. I just added 1 entry. Maybe updating from master did not remove entries that were removed in master?

mrcasals commented:
@tramuntanal after you opened this PR we released `0.13.pre1`, so we dropped the CHANGELOG and restarted it for `master`. Then git gets weird and thinks it should re-add all the contents previous to the line you added.
